File: content/snippets/python/interview_lcci_08_09_bracket.py
from typing import List

def generateParenthesis(n: int) -> List[str]:
    result = []

    def bt(cur, open_count, close_count):
        if len(cur) == 2 * n:
            result.append(cur)
            return
        if open_count < n:
            bt(cur + "(", open_count + 1, close_count)
        if close_count < open_count:
            bt(cur + ")", open_count, close_count + 1)

    bt("", 0, 0)
    return result

def main():
    for s in generateParenthesis(3):
        print(s)

if __name__ == "__main__":
    main()

File: content/snippets/scala3/interview_lcci_08_09_bracket.scala
object Lcci0809Bracket:
    def generateParenthesis(n: Int): Array[String] =
        val result = scala.collection.mutable.ArrayBuffer[String]()

        def bt(cur: String, open: Int, close: Int): Unit =
            if (cur.length == 2 * n) { result += cur; return }
            if (open < n) bt(cur + "(", open + 1, close)
            if (close < open) bt(cur + ")", open, close + 1)

        bt("", 0, 0)
        result.toArray

    @main def main(): Unit =
        generateParenthesis(3).foreach(println)
